include/dlaf/eigensolver/reduction_to_band/impl.h

```diff
@@ -636,64 +614,198 @@ void computePanelReflectors(TriggerSender&& trigger, comm::IndexT_MPI rank_v0,
+          if (tid == 0) {
+            // Note: this is not the final w
+            dlaf::eigensolver::internal::reduceColumnVectors(w_ws);
+            std::copy_n(w_ws[0].data(), pt_cols, algo_data.data() + 2);
+
+            comm::sync::allReduceInPlace(
+                pcomm.get(), MPI_SUM, common::make_data(algo_data.data(), to_SizeType(algo_data_size)));
+          }
 
-          w[index] = common::internal::vector<T>(pt_cols, 0);
-          computeWTrailingPanel(has_head, tiles, w[index], j, pt_cols, begin, end);
           barrier_ptr->arrive_and_wait(barrier_busy_wait);
 
-          // STEP2b: reduce w results (single-threaded)
-          if (index == 0) {
-            dlaf::eigensolver::internal::reduceColumnVectors(w);
-            comm::sync::allReduceInPlace(pcomm.get(), MPI_SUM, common::make_data(w[0].data(), pt_cols));
+          // COMPUTE REFLECTOR and PREPARE PANEL UPDATE
+
+          // Note: if current column is already 0, there's nothing to annihilate.
+          if (algo_data[1] == T(0)) {
+            taus({j, 0}) = 0;
+            continue;
           }
-          barrier_ptr->arrive_and_wait(barrier_busy_wait);
 
-          // STEP3: update trailing panel (multi-threaded)
-          updateTrailingPanel(has_head, tiles, j, w[0], taus({j, 0}), begin, end);
+          const T x0 = algo_data[0];
+          const T norm = std::sqrt(algo_data[1]);
+          const T y = std::signbit(std::real(algo_data[0])) ? norm : -norm;
+          const T tau = (y - x0) / y;
+          const T alpha = T(1) / (x0 - y);
+
+          if (tid == 0) {
+            taus({j, 0}) = tau;
+
+            // Note: correct w* to get the actual w
+            if (pt_cols > 0) {
+              if constexpr (dlaf::isComplex_v<T>)
+                lapack::lacgv(pt_cols, &algo_data[2 + pt_cols], 1);
+              blas::axpy(pt_cols, alpha, &algo_data[2], 1, &algo_data[2 + pt_cols], 1);
+            }
+          }
+
+          // SCAL: compute reflector
+          {
+            bool has_first_component = tid_has_head;
+
+            for (auto i_tl = begin; i_tl < end; ++i_tl) {
+              const auto& tile_v = tiles[i_tl];
+
+              if (has_first_component) {
+                const TileElementIndex idx_x0(j, j);
+                // Note: temporarily keep the reflector well-formed and set band at the end of the iteration
+                tile_v(idx_x0) = 1;
+
+                if (j + 1 < tile_v.size().rows()) {
+                  T* v = tile_v.ptr({j + 1, j});
+                  blas::scal(tile_v.size().rows() - (j + 1), alpha, v, 1);
+                }
+                has_first_component = false;
+              }
+              else {
+                T* v = tile_v.ptr({0, j});
+                blas::scal(tile_v.size().rows(), alpha, v, 1);
+              }
+            }
+          }
+
+          // TODO has pika atomic for bulk threads? condition_variable or atomic?
```

Comment:
@msimberg @biddisco is there anything we can use, which is lighter than a `barrier.wait()`, within bulk?

```cpp
auto ready;
if (tid == 0) {
  axpy(...);
  ready.set();
}
scal(...);
ready.wait();
```

Reply:
Just to record an offline discussion we had a about it:
- condition variable + atomic would be the canonical way to do it
- `pika::event` essentially does the above, packaged into a class
- latch is similar to the above (a tiny bit overfeatured, but close) and it's standard since C++20
- the simplest is an atomic + yield-while (this can be made blocking for a while)

We decided for testing to use the last one, but that can be formalized to a more standard solution if it actually helps.

Reply:
The only place where I see `barrier` being significantly worse is on one gpu for the full eigensolver, and it looks like it might be noise? I would go with the barrier version for now unless there's a strong motivation to keep the more complicated `atomic + yield-while`?

Reply:
I reverted to the basic barrier solution 😉